---
title: "Format UCSC cellbrowser"
author: "Kent Riemondy RBI"
date: "`R Sys.Date()`"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

## Cellbrowser

Format seurat object into a cellbrowser directory suitable for html browser. 

```{r}
library(scbp)
library(tidyverse)
cb_outdir <- "cellbrowser_pub_2022-06"
```

## All cells

```{r}
so <- readRDS(file.path("objects", "sobj.rds"))

ids <- c(2,
         12,
         14,
         16,
         19,
         20,
         24,
         25)

ids <- as.character(ids)

cells <- rownames(so@meta.data)[so@meta.data$v2_refined_clusters %in% ids]
so$cell_lineage <- ifelse(colnames(so) %in% cells,
                            "ameloblast",
                            "other")

cols_to_keep <- c(
  `genes per cell` = "nFeature_RNA",
  `UMIs per cell` = "nCount_RNA",
  `percent of UMIs as mitochondria` = "percent.mt",
  `clusters` = "v2_refined_clusters",
  `cell_lineage` = "cell_lineage",
  `sample` = "orig.ident",
  `cell cycle phase` = "Phase",
  `Tabula Muris cell type` = "tabula_muris_cell_type"
)

make_cellbrowser(so, 
                 column_list = cols_to_keep,
                 secondary_cols = c(
                    "sample"),
                 project = "All-cells",
                 outdir = cb_outdir,
                 marker_file = "markers-v2/all_cells_v2_refined_cluster_markers.tsv",
                 ident = "clusters",
                 embeddings = "umap",
                 skip_expr_matrix = FALSE,
                 description = list(title = "Mouse incisor lineages",
                                    description = "Developing mouse incisor lineages"),
                 cellbrowser_dir = '/Users/kriemo/.local/bin/',
                 )
```

## Just ameloblasts


```{r}
so <- readRDS(file.path("objects", "ablast_sobj_v2.rds"))

cols_to_keep <- c(
  `genes per cell` = "nFeature_RNA",
  `UMIs per cell` = "nCount_RNA",
  `percent of UMIs as mitochondria` = "percent.mt",
  `clusters` = "v2_clusters",
  `sample` = "orig.ident",
  `cell cycle phase` = "Phase"
)


make_cellbrowser(so, 
                 column_list = cols_to_keep,
                 secondary_cols = c(
                    "clusters",
                    "sample"),
                 secondary_color_palette =  c(palette_OkabeIto, rev(discrete_palette_default)),
                 project = "Ameloblasts",
                 outdir = cb_outdir,
                 marker_file = "markers-v2/ablast_subset_v2_cluster_markers.tsv",
                 ident = "clusters",
                 embeddings = "umap",
                 skip_expr_matrix = TRUE,
                 description = list(title = "Ameloblast lineages",
                                    description = "Developing ameloblast lineages"),
                 config = list(priority = 2), 
                 cellbrowser_dir = '/Users/kriemo/.local/bin/',
                 )
```

## Build all

```{r}
library(glue)
write_config <- function(x, path){
  glue('{key}="{value}"', 
     key = names(x), 
     value = x) %>% 
    write_lines(path)
}

collection_cb <- c(shortLabel="Single cell RNA-seq characterization of ameloblast differentiation")

collection_desc <- c(title="Single cell RNA-seq characterization of ameloblast differentiation",
                     abstract="Coordinated mineralization of soft tissue is central to organismal form and function, while dysregulated mineralization underlies several human pathologies. Oral epithelial derived ameloblasts are polarized, secretory cells responsible for generating enamel, the most mineralized substance in the human body. Defects in ameloblast development result in enamel anomalies, including amelogenesis imperfecta. Identifying proteins critical in ameloblast development can provide insight into specific pathologies associated with enamel related disorders or more broadly, mechanisms of mineralization. Previous studies identified a role for MEMO1 in bone mineralization; however, whether MEMO1 functions in the generation of additional mineralized structures remains unknown. Here, we identify a critical role for MEMO1 in enamel mineralization. First, we identified that Memo1 is expressed in ameloblasts and that conditional deletion of Memo1 from ameloblasts results in enamel defects, characterized by a decline in mineral density and tooth integrity. Using the 10x Genomics single-cell RNA-seq platform, we characterized ameloblasts and their progenitors in Memo1 oral epithelial mutants, which revealed a disruption to cytoskeletal associated genes and a reduction in late stage ameloblast markers, relative to controls. ",
                     unitDesc="Log-normalized counts")

write_config(collection_cb, file.path(cb_outdir, "cellbrowser.conf"))
write_config(collection_desc, file.path(cb_outdir, "desc.conf"))

cb_conf <- read_lines("~/.cellbrowser.conf") %>% 
  str_subset("dataRoot") %>% 
  str_split("=") %>% 
  .[[1]] %>% 
  .[2] %>% 
  gsub('\"', "", .)

if(cb_outdir != cb_conf ) {
  write_lines(paste0('dataRoot="', 
                     normalizePath(cb_outdir), '"'),
              "~/.cellbrowser.conf")
}
```

```{r}
cb_dirs <- c(
  "All-cells",
  "Ameloblasts"
)
cb_dirs <- file.path(cb_outdir, 
                     cb_dirs,
                     "cellbrowser.conf")

build_cellbrowser(cb_dirs, 
                  file.path(cb_outdir, "ablast"),
                  cbBuild_path = '/Users/kriemo/.local/bin/cbBuild')
  
```
